send_conn.go

```diff
@@ -13,22 +13,56 @@ type sendConn interface {
 }
 
 type sconn struct {
-	net.PacketConn
+	connection
 
 	remoteAddr net.Addr
+	info       *packetInfo
 }
 
 var _ sendConn = &sconn{}
 
-func newSendConn(c net.PacketConn, remote net.Addr) sendConn {
-	return &sconn{PacketConn: c, remoteAddr: remote}
+func newSendConn(c connection, remote net.Addr, info *packetInfo) sendConn {
+	return &sconn{connection: c, remoteAddr: remote, info: info}
 }
 
 func (c *sconn) Write(p []byte) error {
-	_, err := c.PacketConn.WriteTo(p, c.remoteAddr)
+	_, err := c.WritePacket(p, c.remoteAddr, c.info)
 	return err
 }
 
 func (c *sconn) RemoteAddr() net.Addr {
 	return c.remoteAddr
 }
+
+func (c *sconn) LocalAddr() net.Addr {
+	addr := c.connection.LocalAddr()
+	if c.info != nil {
+		if udpAddr, ok := addr.(*net.UDPAddr); ok {
+			addrCopy := *udpAddr
+			addrCopy.IP = c.info.addr
+			addr = &addrCopy
+		}
+	}
+	return addr
+}
+
+type spconn struct {
```

Comment:
I'm not sure I understand the logic here. As far as I can see, the `spconn` is used by the client, whereas the server would use the `sconn`. That means we assume that the client's kernel will choose the right interface to send the packet from. Is my reasoning correct?

Can we achieve this by using the same struct though? `WritePacket` also works if the `packetInfo` is `nil`, doesn't it?

Reply:
Your reasoning is correct, although, using the same struct for client would require the client code to wrap the `net.PacketConn` to make it a `connection` in all the client code too.